### DISCUSSION TITLE
Bluetooth Mesh stack question

### DISCUSSION THREAD

**Discussion by venngar** [zephyrproject-rtos/zephyr#53927] [General] [answered]
Hello!

I've been developing a complex project involving Bluetooth Mesh API, and I have a question regarding that. In your documentation, you say that Bluetooth API is thread safe unless it's written otherwise:

https://docs.zephyrproject.org/latest/connectivity/bluetooth/bluetooth-dev.html#thread-safety

> Calling into the Bluetooth API is intended to be thread safe, unless otherwise noted in the documentation of the API function. The effort to ensure that this is the case for all API calls is an ongoing one, but the overall goal is formally stated in this paragraph. Bug reports and Pull Requests that move the subsystem in the direction of such goal are welcome.

But my understanding is that it touches only classic Bluetooth API, and not the Mesh. Can you please clarify if Bluetooth Mesh is thread-safe or not?

**Comment by carlescufi** [comment]
@PavelVPV FYI

**Comment by PavelVPV** [comment, marked as answer]
You should only call Bluetooth Mesh API from a cooperative thread. Do not use the system workqueue to call a synchronous Bluetooth Mesh API as it will block the execution of the mesh stack until the API timeout expires. An example of such API will Config Client or Health Client APIs that use `bt_mesh_msg_ack_ctx_match` function (https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/bluetooth/mesh/msg.h#L257-L258) to receive a response in the same execution context. For example, calling `bt_mesh_cfg_cli_app_key_add` with supplied `status` argument from the system workqueue will block sending the Config AppKey Add message until this call times out. In such case, the API will return `ETIMEDOUT` error.

**Reply by venngar** [reply]
Thank your fot the reply!

I see your point and I actually encountered the issue you described with `bt_mesh_cfg_cli_app_key_add` when it tries to self-provision. But isn't system workqueue already a cooperative thread with priority -1? For someone who's just getting into RTOS development, that's an interesting issue to resolve, so I'd like to learn as much as I can. Or is there something I'm missing?

**Reply by PavelVPV** [reply]
Yes, the system workqueue is a cooperative thread. But the mesh stack uses the system workqueue to do process different things: protocol timeouts, messages publication, segments transmission of a segmented message, pushing messages to Bluetooth host, etc. If an application blocks the system workqueue, it will block the stack functionality.

For example, when `bt_mesh_cfg_cli_app_key_add` is called, it should send Config AppKey Add message which is a segmented message. Once an application calls `bt_mesh_cfg_cli_app_key_add` and the complete message is passed from the access layer to the transport layer, the transport layer will split the message to segments, will push the first segment to the mesh advertising bearer and will schedule timeouts. When the extended advertising API is used (which is by default currently), the mesh advertising bearer will schedule a work on the system workqueue to send the message (the first segment) pushed by the transport layer. If an application used `bt_mesh_cfg_cli_app_key_add` in a synchronous manner, it will sleep inside the system workqueue and thus won't let first the mesh advertising bearer to push the message to Bluetooth host. Even with legacy mesh advertising bearer where a separate cooperative thread is used, this will not let the transport layer to schedule the next segment transmission as it also runs on the system workqueue, and etc. This is why an application should not use the system workqueue.

**Reply by venngar** [reply]
Ah, now it's more clear, thank you. In that case, theoretically will the solution of creating a separate workqueue specifically for mesh publishing with the same priority as system workqueue suffice? 

Because, as I understand, making the workqueue with higher priority than system workqueue will block the latter and everything that's scheduled in it.

**Reply by PavelVPV** [reply]
It can have the same priority as the system workqueue, but remember to not run any heavy blocking tasks to block execution of other cooperative threads, which is a usual requirement for the cooperative threads and which also applies to workqueue items (https://docs.zephyrproject.org/latest/kernel/services/scheduling/index.html#cooperative-time-slicing):

> Consequently, if a cooperative thread performs lengthy computations, it may cause an unacceptable delay in the scheduling of other threads, including those of higher priority and equal priority. To overcome such problems, a cooperative thread can voluntarily relinquish the CPU from time to time to permit other threads to execute.

**Reply by venngar** [reply]
Excellent, thank you for the information! This will get useful

**Reply by PavelVPV** [reply]
@venngar, could you please mark the question as answered?

**Reply by venngar** [reply]
Done, and thanks once again!
